tests/test_java.py

```diff
@@ -60,9 +69,9 @@ def test_async_profiler_already_running(application_pid, assert_collapsed, tmp_p
             ap_safemode=0,
             ap_args="",
         ) as ap_proc:
-            ap_proc.status_async_profiler()
+            ap_proc.status_async_profiler()  # type: ignore
```

Comment:
It's our class :/ why `type ignore`

Reply:
Glad you asked :)

The function `status_async_profiler` isn't defined in `AsyncProfiledProcess`, it's defined in `AsyncProfiledProcessForTests` that inherits from `AsyncProfiledProcess`, and `AsyncProfiledProcess::_enter__()` returns `AsyncProfiledProcess`.

I didn't want to make `AsyncProfiledProcess` generic because I felt it's a hack just got testing, but now I think we should.

Changed `__enter__()`'s return type to be `T` and removed `type: ignore`s

Reply:
> The function status_async_profiler isn't defined in AsyncProfiledProcess, it's defined in AsyncProfiledProcessForTests that inherits from AsyncProfiledProcess, and AsyncProfiledProcess::_enter__() returns AsyncProfiledProcess.

Hmmm. I wonder what is the correct solution here.

Reply:
You can't have the cake and eat it too.

`__enter__()` has to declare it returns the class (instead of the hard-coded `AsyncProfiledProcess`), and if we don't want that, then we shouldn't inherit from it and add functions.  
`AsyncProfiledProcess` can either support inheritance, or not, and in that case, don't inherit from it (tests shouldn't be an exception to that)

So I chose to support inheritance and treat the testing code as a legitimate use-case

Reply:
I'm satisfied with the solution here, per my understanding from https://stackoverflow.com/a/59937840/797390